### PR TITLE
Feature/35224 email spec service

### PR DESCRIPTION
**Pull request by YRedko**

### CODE REVIEW

File: src/main/java/com/icthh/xm/tmf/ms/communication/domain/spec/EmailTemplateSpec.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.experimental.Accessors;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@Accessors(chain = true)
+public class EmailTemplateSpec implements Comparable<EmailTemplateSpec>{
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/communication/domain/spec/EmailTemplateSpec.java:12]: formatting

File: src/main/java/com/icthh/xm/tmf/ms/communication/domain/spec/EmailTemplateSpec.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    private String templateKey;
+    private String name;
+    private String defaultSubjectTemplate;
+    private String templatePath;
+    private String defaultContext;
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/communication/domain/spec/EmailTemplateSpec.java:17]: maybe better to contextExample

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecificationService.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class EmailSpecificationService implements RefreshableConfiguration {
+
+    private static final String TENANT_NAME = "tenantName";
+
+    private ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
+    private final ConcurrentHashMap<String, Map<String, EmailTemplateSpec>> emailSpecs = new ConcurrentHashMap<>();
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecificationService.java:35]: on left side of declaration use interface only

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecificationService.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    private final ConcurrentHashMap<String, Map<String, EmailTemplateSpec>> customerEmailSpecs = new ConcurrentHashMap<>();
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecificationService.java:36]: on left side ConcurrentHashMap -> Map

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecificationService.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    private final TenantContextHolder tenantContextHolder;
+
+    @LoggingAspectConfig(resultDetails = false)
+    public List<EmailTemplateSpec> getEmailSpecList() {
+        String tenantKey = tenantContextHolder.getTenantKey();
+        String cfgTenantKey = tenantKey.toUpperCase();
+        if (!emailSpecs.containsKey(cfgTenantKey)) {
+            throw new IllegalArgumentException("Email specification not found");
+        }
+        Map<String, EmailTemplateSpec> emailSpec = Optional.ofNullable(emailSpecs.get(cfgTenantKey)).orElseGet(Collections::emptyMap);
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecificationService.java:48]: emailSpecs.getOrDefault(cfgTenantKey, emptyMap())

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecificationService.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        Map<String, EmailTemplateSpec> customerEmailSpec = Optional.ofNullable(customerEmailSpecs.get(cfgTenantKey)).orElseGet(Collections::emptyMap);
+        customerEmailSpec.keySet().retainAll(emailSpec.keySet());
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecificationService.java:50]: modification state of service on read? O_o

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecificationService.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        processedEmailSpec.putAll(customerEmailSpec);
+
+        return new ArrayList<>(processedEmailSpec.values());
+    }
+
+    @Override
+    public void onRefresh(String updatedKey, String config) {
+        try {
+            String tenantName = getTenantKey(updatedKey);
+            ConcurrentHashMap<String, Map<String, EmailTemplateSpec>> updatedSpec = getUpdatedSpec(updatedKey);
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecificationService.java:61]: implementation on declaration =(

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecificationService.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            log.error("Error read specification from path {}", updatedKey, e);
+        }
+    }
+
+    @Override
+    public boolean isListeningConfiguration(String updatedKey) {
+        return matcher.match(properties.getEmailSpecificationPathPattern(), updatedKey);
+    }
+
+    private ConcurrentHashMap<String, Map<String, EmailTemplateSpec>> getUpdatedSpec(String updatedKey) {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecificationService.java:84]: Map

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecificationService.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        if (matcher.match(properties.getEmailSpecificationPathPattern(), updatedKey)) {
+            return emailSpecs;
+        } else {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecificationService.java:87]: else if match another pattern or emptyMap

File: src/main/java/com/icthh/xm/tmf/ms/communication/domain/spec/EmailTemplateSpec.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.experimental.Accessors;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@Accessors(chain = true)
+public class EmailTemplateSpec implements Comparable<EmailTemplateSpec> {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/communication/domain/spec/EmailTemplateSpec.java:12]: abstract

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecificationService.java [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+    private ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
+    private final Map<String, Map<String, EmailTemplateSpec>> emailSpecs = new ConcurrentHashMap<>();
+    private final Map<String, Map<String, EmailTemplateSpec>> customerEmailSpecs = new ConcurrentHashMap<>();
+    private final ApplicationProperties properties;
+    private final AntPathMatcher matcher = new AntPathMatcher();
+    private final TenantContextHolder tenantContextHolder;
+
+    @LoggingAspectConfig(resultDetails = false)
+    public List<EmailTemplateSpec> getEmailSpecList() {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/communication/service/EmailSpecificationService.java:47]: DefaultEmailTemplateSpec

**Review by sergeysenja1992 — APPROVED** [review]

**Review by monassfaro — APPROVED** [review]

**Review by sergeysenja1992 — APPROVED** [review]

**Pull request merged**